src/wasm-interpreter.h

```diff
@@ -1541,13 +1543,37 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
 
 private:
   // Truncate the value if we need to. The storage is just a list of Literals,
-  // so we can't just write the value like we would to a C struct field.
-  Literal getMaybePackedValue(Literal value, const Field& field) {
+  // so we can't just write the value like we would to a C struct field and
+  // expect it to truncate for us. Instead, we truncate so the stored value is
+  // proper for the type.
+  Literal truncateForPacking(Literal value, const Field& field) {
+    if (field.type == Type::i32) {
+      int32_t c = value.geti32();
+      if (field.packedType == Field::i8) {
+        value = Literal(c & 0xff);
+      } else if (field.packedType == Field::i16) {
+        value = Literal(c & 0xffff);
+      }
+    }
+    return value;
+  }
+
+  Literal extendForPacking(Literal value, const Field& field, bool signed_) {
     if (field.type == Type::i32) {
+      int32_t c = value.geti32();
       if (field.packedType == Field::i8) {
-        value = value.and_(Literal(int32_t(0xff)));
+        if (signed_) {
+          value = Literal((c << 24) >> 24);
+        } else {
+          // The stored value should already be truncated.
+          assert(c == (c & 0xff));
```

Comment:
This should be true in both cases, so how about moving this just before the `if (signed_)`? Same below.

Reply:
Good idea, that's shorter too. Done.